### ISSUE TITLE
Configuration for LLaMA

### ISSUE THREAD

**Issue by bangawayoo** [lancopku/codable-watermarking-for-llm#3]
Hi.
Thank you for your work!

I ran the experiment for OPT, which works fine after changing how the model is loaded.
(I am using `AutoModelForCausalLM.from_pretrained(...)`.)

However, I am having some trouble when trying to experiment on LLaMA-7b.
The accuracy is near 20% even when the settings are modified to make the task easier:

> generated_length: int = 100
    message_code_len: int = 4
    encode_ratio: float = 25.

This seems like something is not done properly on during generation.

Do I have to change the setting for LLaMA?
Could you share the configuration for running on LLaMA? 

Any help would be appreciated.
Thanks!

**Comment by bangawayoo** [comment]
The reason behind this was the difference in the tokenizers during encoding and decoding of messages. I verified this by testing with LLaMA-2-7b in the Vanilla Marking framework. I did have to pass the original tokenizer to the `RandomMessageModel`, which seems like hotfix

> RandomMessageModel(tokenizer=tokenizer,
>                                               lm_tokenizer=tokenizer,
>                                               delta=args.delta,
>                                               message_code_len=args.message_code_len,
>                                               device=model.device,
>                                               )


I am still working out how to fix this for Balance Marking. So, any help would be appreciated!

It seems like this should be handled by converting through `self.tokenizer_id_2_lm_tokenizer_id_list`. I am not sure why this is not working. 

The intuition behind this is to **hash with `lm_tokenizer` (e.g. GPT-2) not the original tokenizer** both during generation and decoding, right?

----
Update: Changing the `lm_tokenizer` for `LMMessageModel` seems to work. Lean, do you think does Balance Marking as intended? 

Thanks!

**Comment by leanwang326** [comment]
Sorry for replying so late. Yes, that's a good question. 
The question is, you can not convert through self.tokenizer_id_2_lm_tokenizer_id_list when using llama tokenizer(llama's token has ""). llama's tokenizer has many different subtokens with gpt/opt. For example, the ' A' in gpt/opt is 'ĠA', while in llama is '▁A'. A quick fix is using 

```
  def construct_tokenizer_id_2_lm_tokenizer_id_list(self):
    tokenizer_id_2_lm_tokenizer_id_list = torch.zeros(len(self.tokenizer.vocab),
                                                              dtype=torch.long)
    for v, i in self.tokenizer.vocab.items():
        if 'llama' in self.tokenizer.name_or_path and 'gpt' in self.lm_tokenizer.name_or_path:
            # TODO: may need further fix
            v = v.replace('▁', 'Ġ')
        tokenizer_id_2_lm_tokenizer_id_list[i] = self.lm_tokenizer.convert_tokens_to_ids(v)
    return tokenizer_id_2_lm_tokenizer_id_list.to(self.lm_model.device) 
```
but the result is still not so good, since llama usually has shorter subtokens. This may require more complex technical solution, like combining subtokens and aligning the logits in word level / using word level hash, which we haven't developed one yet. Anyway, I may suggest using the proxy-LM with the same tokenizers (e.g. gpt-small for gpt-xl, llama-7b(or its quantized version) for llama-13b) for now.

**Comment by bangawayoo** [comment]
Thanks for the reply.
Now I understand why the problem consists after `self.tokenizer_id_2_lm_tokenizer_id_list`.
